x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/DelegatedAuthorizationSupport.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.authc.support;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.xpack.core.common.IteratingActionListener;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationResult;
+import org.elasticsearch.xpack.core.security.authc.Realm;
+import org.elasticsearch.xpack.core.security.authc.RealmConfig;
+import org.elasticsearch.xpack.core.security.authc.support.DelegatedAuthorizationSettings;
+import org.elasticsearch.xpack.core.security.user.User;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.elasticsearch.common.Strings.collectionToDelimitedString;
+
+public class DelegatedAuthorizationSupport {
+
+    private final List<Realm> lookupRealms;
+    private final ThreadContext threadContext;
+
+    public DelegatedAuthorizationSupport(Iterable<? extends Realm> allRealms, RealmConfig config) {
+        this(allRealms, DelegatedAuthorizationSettings.AUTHZ_REALMS.get(config.settings()), config.threadContext());
+    }
+
+    protected DelegatedAuthorizationSupport(Iterable<? extends Realm> allRealms, List<String> lookupRealms, ThreadContext threadContext) {
+        this.lookupRealms = resolveRealms(allRealms, lookupRealms);
+        this.threadContext = threadContext;
+    }
+
+    public boolean hasDelegation() {
+        return this.lookupRealms.isEmpty() == false;
+    }
+
+    public void resolveUser(String username, ActionListener<AuthenticationResult> resultListener) {
+        if (lookupRealms.isEmpty()) {
+            throw new IllegalStateException("No realms have been configured for delegation");
```

Comment:
Really? Because the method doesn't actually return anything, so it would need to call `onResponse` with an Optional, and you wouldn't necessarily know what the _absent_ value represented.
Does it mean no realms configured, or does it mean the user couldn't be found?

Maybe you could show me how it might work, because I'm not seeing it.

Contractually, I think it is an error to call this if there are no lookup realms. I could see an argument for passing the exception through `onFailure` instead, but if this method is called without any lookup realms then the calling code (the originating realm) has a bug, because it should check `hasDelegation()` first.

I did originally have an interface where the "if no lookup" body was passed in as a `Function` so this method never threw an exception, but it made the calling code messier, with no benefit.